source/AI.cpp

```diff
@@ -3313,32 +3313,103 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
 		// Board the nearest disabled ship, focusing on hostiles before allies. Holding
 		// `Shift` results in boarding only player-owned escorts in need of assistance.
 		shared_ptr<const Ship> target = ship.GetTargetShip();
-		if(!target || !CanBoard(ship, *target) || (shift && !target->IsYours()))
+		if(!target || (target && activeCommands.Has(Command::WAIT)) || !CanBoard(ship, *target) || (shift && !target->IsYours()))
 		{
 			if(shift)
 				ship.SetTargetShip(shared_ptr<Ship>());
 
-			double closest = numeric_limits<double>::infinity();
 			bool foundEnemy = false;
-			bool foundAnything = false;
-			for(const shared_ptr<Ship> &other : ships)
-				if(CanBoard(ship, *other))
+			bool boardClosest = Preferences::Has("Board target");
+
+			auto strategy = [&]() -> function<double(const Ship &)>
+			{
+				Point current = ship.Position();
+				// If "board target" is set to false, value priority is used.
+				if(!boardClosest)
 				{
-					if(shift && !other->IsYours())
-						continue;
+					return [this, &ship](const Ship &other) -> double
+					{
+						// Use the exact cost if the ship was scanned, otherwise use an estimation.
+						return this->Has(ship, other.shared_from_this(), ShipEvent::SCAN_OUTFITS) ?
+							other.Cost() : (other.ChassisCost() * 2.);
+					};
+				}
+				// Default to distance-based strategy.
+				return [current](const Ship &other) -> double
+				{
+					return current.DistanceSquared(other.Position());
+				};
+			}();
+
+			auto boardable = vector<pair<const Ship *, double>>{};
+
+			auto fillBoardable = [&ship, &foundEnemy, &boardable, &strategy](const Ship &other)
+			{
+				if(CanBoard(ship, other))
+					boardable.emplace_back(make_pair(&other, strategy(other)));
+			};
+
+			if(shift)
+			{
+				for(const Ship *other : governmentRosters[ship.GetGovernment()])
+					if(other->GetSystem() == ship.GetSystem())
+						fillBoardable(*other);
+			}
+			else
+			{
+				// First check if we can board enemy ships, then allies.
+				for(const Ship *other : GetShipsList(ship, true))
+					fillBoardable(*other);
+				// The current target is not considered by GetShipsList.
+				if(target)
+					fillBoardable(*target.get());
+
+				if(boardable.empty())
+					for(const Ship *other : GetShipsList(ship, false))
+						fillBoardable(*other);
+			}
 
-					bool isEnemy = other->GetGovernment()->IsEnemy(ship.GetGovernment());
-					double d = other->Position().Distance(ship.Position());
-					if((isEnemy && !foundEnemy) || (d < closest && isEnemy == foundEnemy))
+			if(boardable.empty())
+				activeCommands.Clear(Command::BOARD);
+			else
+			{
+				sort(boardable.begin(), boardable.end(),
+					[&ship, boardClosest](
+						pair<const Ship *, double> &lhs,
+						pair<const Ship *, double> &rhs
```

Comment:
Sorting should not modify the sorted elements
```suggestion
						const pair<const Ship *, double> &lhs,
						const pair<const Ship *, double> &rhs
```
A function-level `using` directive could be helpful to provide a shorter type alias, e.g. "TargetPriority"